File: lib/core/device_provider.dart
// ignore_for_file: close_sinks

import 'dart:io';
import 'dart:typed_data';

import 'package:unisync/utils/logger.dart';

import '../models/device_info/device_info.model.dart';
import 'device.dart';
import 'device_connection.dart';

class DeviceProvider {
  DeviceProvider._();

  static final _devices = <DeviceInfo>{};

  static List<DeviceInfo> get connectedDevices => _devices.toList();

  static void create({required DeviceInfo info, required SecureSocket socket, required Stream<Uint8List> inputStream}) {
    if (_devices.contains(info)) {
      infoLog('DeviceProvider: Duplicate connection to ${info.name}. Disconnecting...');
      socket.close();
    } else {
      _devices.add(info);
      final connection = DeviceConnection(socket, inputStream, () {
        _devices.remove(info);
      });
      Device(info).connection = connection;
    }
  }
}

File: unisync_android/app/src/main/java/com/anhquan/unisync/core/DeviceProvider.kt
package com.anhquan.unisync.core

import android.content.Context
import com.anhquan.unisync.models.DeviceInfo
import com.anhquan.unisync.utils.infoLog
import javax.net.ssl.SSLSocket

object DeviceProvider {
    private val _devices = mutableSetOf<DeviceInfo>()

    fun create(
        context: Context,
        info: DeviceInfo,
        socket: SSLSocket
    ) {
        if (_devices.contains(info)) {
            infoLog("${this::class.simpleName}: Duplicate connection to ${info.name}. Disconnecting...")
            socket.close()
        } else {
            _devices.add(info)
            val connection = DeviceConnection(
                context,
                socket,
            ) {
                _devices.remove(info)
            }
            Device.of(info).apply {
                this.connection = connection
            }
        }
    }
}